src/auditing.coffee

```diff
@@ -0,0 +1,72 @@
+logger = require 'winston'
+syslogParser = require('glossy').Parse
+parseString = require('xml2js').parseString
+firstCharLowerCase = require('xml2js').processors.firstCharLowerCase
+Audit = require('./model/audits').Audit
+
+parseAuditRecordFromXML = (xml, callback) ->
+  # DICOM mappers
+  csdCodeToCode = (name) -> if name is 'csd-code' then 'code' else name
+  originalTextToDisplayName = (name) -> if name is 'originalText' then 'displayName' else name
+
+  options =
+    mergeAttrs: true,
+    explicitArray: false
+    tagNameProcessors: [firstCharLowerCase]
+    attrNameProcessors: [firstCharLowerCase, csdCodeToCode, originalTextToDisplayName]
+
+  parseString xml, options, (err, result) ->
+    return callback err if err
+
+    if not result?.auditMessage
+      return callback 'Document is not a valid AuditMessage'
+
+    audit = {}
+
+    if result.auditMessage.eventIdentification
+      audit.eventIdentification = result.auditMessage.eventIdentification
+
+    audit.activeParticipant = []
+    if result.auditMessage.activeParticipant
+      # xml2js will only use an array if multiple items exist (explicitArray: false), else it's an object
+      if result.auditMessage.activeParticipant instanceof Array
+        for ap in result.auditMessage.activeParticipant
+          audit.activeParticipant.push ap
+      else
+        audit.activeParticipant.push result.auditMessage.activeParticipant
+
+    if result.auditMessage.auditSourceIdentification
+      audit.auditSourceIdentification = result.auditMessage.auditSourceIdentification
+
+    audit.participantObjectIdentification = []
+    if result.auditMessage.participantObjectIdentification
+      # xml2js will only use an array if multiple items exist (explicitArray: false), else it's an object
+      if result.auditMessage.participantObjectIdentification instanceof Array
+        for poi in result.auditMessage.participantObjectIdentification
+          audit.participantObjectIdentification.push poi
+      else
+        audit.participantObjectIdentification.push result.auditMessage.participantObjectIdentification
+
+    callback null, audit
+
+
+exports.processAudit = (msg, callback) ->
+  parsedMsg = syslogParser.parse(msg)
+
+  if not parsedMsg or not parsedMsg.message
+    logger.info 'Invalid message received'
+    return callback()
+
+  parseAuditRecordFromXML parsedMsg.message, (err, result) ->
+    audit = new Audit result
+
+    audit.rawMessage = msg
+    audit.syslog = parsedMsg
+    delete audit.syslog.originalMessage
+    delete audit.syslog.message
+
+    audit.save (saveErr) ->
+      if err or saveErr
```

Comment:
The use case I have in mind is, esp. with the UDP interface, if messages come through incomplete. That way we still store the raw message, even if we can't parse it. I saw this with a few times with OpenATNA and those big PnR messages. If we do get to the `err` stage, it also means it's valid syslog, so we won't be auditing garbage.

What do you think?

FYI looking at this code again, i see it's not too readable, just made a couple of small edits...